src/circuit/gadget/utilities/decompose_running_sum.rs

```diff
@@ -0,0 +1,369 @@
+    };
+    use pasta_curves::{arithmetic::FieldExt, pallas};
+
+    #[test]
+    fn test_running_sum() {
+        struct MyCircuit<
+            F: FieldExt + PrimeFieldBits,
+            const WORD_NUM_BITS: usize,
+            const WINDOW_NUM_BITS: usize,
+            const NUM_WINDOWS: usize,
+        > {
+            alpha: Option<F>,
+            strict: bool,
+        }
+
+        impl<
+                F: FieldExt + PrimeFieldBits,
+                const WORD_NUM_BITS: usize,
+                const WINDOW_NUM_BITS: usize,
+                const NUM_WINDOWS: usize,
+            > Circuit<F> for MyCircuit<F, WORD_NUM_BITS, WINDOW_NUM_BITS, NUM_WINDOWS>
+        {
+            type Config = RunningSumConfig<F, WORD_NUM_BITS, WINDOW_NUM_BITS, NUM_WINDOWS>;
+            type FloorPlanner = SimpleFloorPlanner;
+
+            fn without_witnesses(&self) -> Self {
+                Self {
+                    alpha: None,
+                    strict: self.strict,
+                }
+            }
+
+            fn configure(meta: &mut ConstraintSystem<F>) -> Self::Config {
+                let z = meta.advice_column();
+                let q_range_check = meta.selector();
+                let perm = meta.permutation(&[z.into()]);
+
+                RunningSumConfig::<F, WORD_NUM_BITS, WINDOW_NUM_BITS, NUM_WINDOWS>::configure(
+                    meta,
+                    q_range_check,
+                    z,
+                    perm,
+                )
+            }
+
+            fn synthesize(
+                &self,
+                config: Self::Config,
+                mut layouter: impl Layouter<F>,
+            ) -> Result<(), Error> {
+                layouter.assign_region(
+                    || "decompose",
+                    |mut region| {
+                        let offset = 0;
+                        let (alpha, _zs) = config.witness_decompose(
+                            &mut region,
+                            offset,
+                            self.alpha,
+                            self.strict,
+                        )?;
+
+                        let offset = offset + NUM_WINDOWS + 1;
+
+                        config.copy_decompose(&mut region, offset, alpha, self.strict)?;
+
+                        Ok(())
+                    },
+                )
+            }
+        }
+
+        // Random base field element
+        {
+            let alpha = pallas::Base::rand();
+
+            // Strict full decomposition should pass.
+            let circuit: MyCircuit<
+                pallas::Base,
+                L_ORCHARD_BASE,
+                FIXED_BASE_WINDOW_SIZE,
+                NUM_WINDOWS,
+            > = MyCircuit {
+                alpha: Some(alpha),
+                strict: true,
+            };
+            let prover = MockProver::<pallas::Base>::run(8, &circuit, vec![]).unwrap();
+            assert_eq!(prover.verify(), Ok(()));
+
+            // Strict partial decomposition should fail.
```

Comment:
Is this true for all possible alpha, or only true except with negligible probability?

Reply:
I am now only testing the partial decomposition failure cases using `alpha` strictly larger than 64 bits.